proposals/csharp-11.0/low-level-struct-improvements.md

```diff
@@ -869,6 +869,149 @@ ref struct S<T>
+- `lifetime $this >= $ro`
+
+Note: while both `$this` and `$cm` are `>= $ro` there is no relationship between the two of them.
+
+The C# method syntax maps to the model in the following ways: 
+
+- `ref` parameters have a ref lifetime of `$ro`
+- parameters of type `ref struct` have a value lifetime of `$cm`
+- ref returns have a ref lifetime of `$ro`
+- returns of type `ref struct` have a value lifetime of `$ro`
+- `scoped` on a parameter or `ref` changes the lifetime to be `$local`
+- `ref this` has lifetime `$local` by default but it becomes `$ro` when the method is annotated with `[UnscopedRef]`
+
+The basic rules for the lifetime are defined as:
+
+- All lifetimes are expressed syntactically as generic arguments, coming before type arguments. This is true for predefined lifetimes except `$heap` and `$local`. 
+- All variables have their lifetime defined at declaration by annotating the type. For example `$a ref $b Span<int> span` defines a local `span` whose value lifetime is `$b` and whose ref lifetime is `$a`. 
+- All values that are not typed to `ref struct` implicitly have a lifetime of `$heap`. That is there is no need to write `$heap int` everywhere, one can simply write `$int`. 
+- The `ref` of a non-ref local has ref lifetime `$local`
+- An assignment is only legal when the lifetime of the RHS is greater than or equal to the LHS.
+- An return is only legal when the lifetime of the value is greater than or equal to lifetime on the return
+- Lifetimes of expressions can be made explicit by putting annotations in front of them:
+    - `$a expr` the value lifetime is explicitly `$a`
+    - `$a ref $b expr` the value lifetime is `$b` and the ref lifetime is `$a`.
+- A method call has the value return of the smallest of all the lifetimes input: ref or value. In the case there is no total ordering of the lifetimes, for example if two lifetimes have no relationship to each other, then the result is a new named lifetime that has no relationship to any other lifetime.
+
+Given that let's explore a simple example that demonstrates the model here: 
+
+```csharp
+ref int M1(ref int i) => ...
+
+// Maps to the following. 
+
+$ro ref int Identity<$ro>($ro ref int i)
+{
+    // okay: has ref lifteime $ro which is equal to $ro
+    return ref i;
+
+    // okay: has ref lifetime $heap which is >= $ro
+    int[] array = new int[42];
+    return ref array[0];
+
+    // error: has ref lifetime $local which has no relationship to $a hence 
+    // it's illegal
+    int local = 42;
+    return ref local;
+}
+```
+
+Now let's explore the same example using a `ref struct`: 
+
+```csharp
+ref struct S
+{
+    ref int Field;
+
+    $ro S<$ro>($ro ref int f)
+    {
+        Field = ref f;
+    }
+}
+
+S M2(ref int i, S span1, scoped S span2) => ...
+
+// Maps to 
+
+$ro S<int> M2<$ro>(
+    $ro ref int i,
+    $ro S span1)
+    $local S span2)
```

Comment:
>  it can't be the same lifetime as local ref variables.

Consider that [this ](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEUCuA7AHwAEAmARgFgAoaosgTgAoAiACxgBsOJmBKAbmq0UAAgCyjAM6QADjAAmIgMoyAhngA8ASzwYAfCMlq8vEQG9qIq8uPbdByRlVgA1ivUiAvIaevVXCDARHQwAbRQSAF1BKmtDYy8fZzdjGIBfIA===)is completely legal today: 

```csharp
void M(scoped Span<int> span) {
    Span<int> stackSpan = stackalloc int[42];
    span = stackSpan;
}
```

This is why I defined local in this way

Reply:
I stand corrected! This is insane, but it *is* what the language implemented.